cnqr/_src/convex.py

```diff
@@ -79,103 +106,104 @@ def __call__(self, x, y):
+            DenseWrapper,
+            layer_class=PositiveDense,
+            kernel_init_type='constant',
+            use_bias=False,
+        )
         rescale = PositiveDense.inv_rectifier_fn
-        for i in range(0, num_hidden + 1):
-            output_dim = 1 if i == num_hidden else self.dim_hidden[i]
-            w_U.append(
-                PositiveDense(
-                    output_dim,
-                    name=f'w_U_pdense_{i}',
-                    kernel_init=initializers.constant(
-                        rescale(1.0 / normalization)),
-                    use_bias=False,
-                )
-            )
-            if i < num_hidden:
-                normalization = self.dim_hidden[i]
-
-        w_V = []  # dense layers (no bias), init: weights -> approx(0)
-        for i in range(0, num_hidden + 1):
-            output_dim = 1 if i == num_hidden else self.dim_hidden[i]
-            w_V.append(
-                nn.Dense(
-                    output_dim,
-                    name=f'w_V_dense_{i}',
-                    kernel_init=self.init_fn(self.init_std),
-                    use_bias=False,
-                )
-            )
+        kernel_constant_inits_U = [rescale(1.0 / d1) for d1 in [1] + list(self.dim_hidden)]
+        output_dims_U = list(self.dim_hidden) + [1]
+
+        # w_V: dense layers (no bias), weights init -> approx(0)
+        w_V = partial(
+            DenseWrapper,
+            kernel_init_type='normal',
+            kernel_epsilon_init=self.epsilon_init,
+            use_bias=False,
+        )
+        output_dims_V = list(self.dim_hidden) + [1]
 
         # ===== Inference =====
         x = phi(x)
-        z = q(y) * jnp.ones(shape=x.shape)
-        for i in range(num_hidden + 1):
+        z = q(y).reshape(x.shape[0], -1)
+        num_steps = len(self.dim_hidden) + 1
+        for i in range(num_steps):
             # context modulators
-            u = f['f_u'][i](x)
-            v = f['f_v'][i](x)
-            w = f['f_w'][i](x)
+            u = f_u(features=output_dims_u[i], name=f'f_u_dense_{i}')(x)
+            v = f_v(features=output_dims_v[i], name=f'f_v_dense_{i}')(x)
+            w = f_w(features=output_dims_w[i], name=f'f_w_dense_{i}')(x)
             # convex network
-            h = w_U[i](z * nn.softplus(u)) + w_V[i](y * v) + w
-            z = self.act_fn_z(h)
+            w_Uk = w_U(features=output_dims_U[i], name=f'w_U_pdense_{i}',
+                       kernel_constant_init=kernel_constant_inits_U[i])
+            w_Vk = w_V(features=output_dims_V[i], name=f'w_V_dense_{i}')
+            h = w_Uk(z * self.pos_act_fn(u)) + w_Vk(y * v) + w
+            z = self.sigma_act_fn(h)
             # context network
-            if i < num_hidden:
-                x = self.act_fn_x(f_x[i](x))
+            if i < num_steps - 1:
+                x = f_x(features=output_dims_x[i], name=f'f_x_dense_{i}')(x)
+                x = self.tau_act_fn(x)
         return jnp.squeeze(z)
```

Comment:
A bit dangerous IMO: if the batch size is equal to 1, the corresponding dimension will be removed, which may cause issues since the output shape is no longer `(batch_size,)` as expected. It is better to use a good old' fashion `reshape` here.

Reply:
True, I changed the return into `x.reshape(-1,)` and adapted the code in the convex_test.py file accordingly, thanks for the review!